File: roll-your-own-test-framework/python/tests/test_suite_builder.py
from typing import Type

from roll_your_own_test_framework import AssertionFailedException


class TestSuiteBuilder:
    def __init__(self) -> None:
        self._methods: dict[str, object] = {}

    def with_passing_test(self, name: str = "test_passing") -> "TestSuiteBuilder":
        def passing(self_inner: object) -> None:
            pass

        self._methods[name] = passing
        return self

    def with_failing_test(
        self, name: str = "test_failing", message: str = "expected 5 but got 3"
    ) -> "TestSuiteBuilder":
        def failing(self_inner: object) -> None:
            raise AssertionFailedException(message)

        self._methods[name] = failing
        return self

    def with_erroring_test(
        self, name: str = "test_erroring", message: str = "something went wrong"
    ) -> "TestSuiteBuilder":
        def erroring(self_inner: object) -> None:
            raise RuntimeError(message)

        self._methods[name] = erroring
        return self

    def build(self) -> Type:
        return type("DynamicTestSuite", (), dict(self._methods))

File: roll-your-own-test-framework/typescript/tests/TestSuiteBuilder.ts
import { AssertionFailedException } from '../src/AssertionFailedException.js';
import { TestSuite } from '../src/TestRunner.js';

export class TestSuiteBuilder {
  private readonly tests: Record<string, () => void> = {};

  withPassingTest(name: string = 'passingTest'): this {
    this.tests[name] = () => {};
    return this;
  }

  withFailingTest(name: string = 'failingTest', message: string = 'expected 5 but got 3'): this {
    this.tests[name] = () => {
      throw new AssertionFailedException(message);
    };
    return this;
  }

  withErroringTest(name: string = 'erroringTest', message: string = 'something went wrong'): this {
    this.tests[name] = () => {
      throw new Error(message);
    };
    return this;
  }

  build(): TestSuite {
    return { ...this.tests };
  }
}
